lib/Sema/SemaType.cpp

```diff
@@ -4252,20 +4293,23 @@ static TypeSourceInfo *GetFullTypeForDeclarator(TypeProcessingState &state,
         checkNullabilityConsistency(S, SimplePointerKind::Array, DeclType.Loc);
       }
 
-      bool isChecked = ATI.isChecked;
+      CheckedArrayKind Kind = (CheckedArrayKind) ATI.kind;
       // Handle multi-dimensional arrays for Checked C. A multi-dimensional
       // array is an array of arrays, so look for a nested type that is an array.
       // - Dimensions must either all be checked or all be unchecked.
       // - The checked property propagates to nested array types, if the array
       //   types are unchecked.  The nested array types must be declared as part
       //   of this declaration.
       if (const ArrayType *AT = dyn_cast<ArrayType>(T.getCanonicalType())) {
-        if (isChecked != AT->isChecked()) {
-          if (isChecked) 
-            // The new array type is checked. Propagate this to nested array types
-            // declared as part of this declaration.
+        if (Kind != AT->getKind()) {
+          if (Kind == CheckedArrayKind::Checked)
+            // The new array type is checked. Propagate this to nested array
+            // types declared as part of this declaration.
             T = S.MakeCheckedArrayType(T, true, DeclType.Loc);
-          else {
+          else if (Kind == CheckedArrayKind::NtChecked) {
+            // Do not propagate - null-terminated arrays of arrays are not
+            // allowed and BuildArrayType will issue an error.
```

Comment:
Does that mean this isn't reachable? Should there be an `llvm_unreachable` here? Or the error will happen later?